atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -205,42 +229,33 @@ private void debugLogStateOfPool() {
 
     private void checkAndUpdateBlacklist() {
         // Check blacklist and re-integrate or continue to wait as necessary
-        for (Map.Entry<InetSocketAddress, Long> blacklistedEntry : blacklistedHosts.entrySet()) {
+        for (Map.Entry<InetSocketAddress, Long> blacklistedEntry : unresponsiveHostBlacklist.getHosts().entrySet()) {
             long backoffTimeMillis = TimeUnit.SECONDS.toMillis(config.unresponsiveHostBackoffTimeSeconds());
             if (blacklistedEntry.getValue() + backoffTimeMillis < System.currentTimeMillis()) {
                 InetSocketAddress host = blacklistedEntry.getKey();
                 if (isHostHealthy(host)) {
-                    blacklistedHosts.remove(host);
+                    unresponsiveHostBlacklist.removeHost(host);
                     log.error("Added host {} back into the pool after a waiting period and successful health check.", host);
                 }
             }
         }
     }
 
-    private void addToBlacklist(InetSocketAddress badHost) {
-        blacklistedHosts.put(badHost, System.currentTimeMillis());
-    }
-
     private boolean isHostHealthy(InetSocketAddress host) {
-        CassandraClientPoolingContainer testingContainer = null;
         try {
-            testingContainer = new CassandraClientPoolingContainer(host, config);
-            testingContainer.runWithPooledResource(describeRing);
+            currentPools.get(host).runWithPooledResource(describeRing);
```

Comment:
I discussed the race in the commit message:

> Don't create a separate cassandra pool to healthcheck cassandra hosts
> 
> Extra pools cause noise with jmx metrics (each pool has its own set of
> metrics which aren't removed from the MBean server on shutdown).
> 
> This is also unnecessary since we should have a pool for each host.
> Technically, the existing code did not maintain that invariant (which
> implicitly resulted in it healthchecking blacklisted hosts that were no
> longer in the config). Maintaining that invariant adds a small
> complication (we _must_ remove a host from blacklistedHosts before
> removing the corresponding pool from currentPools) but it doesn't seem
> too bad

IIRC the CassandraPoolingContainer throws away connections that were bad in some way and I think it's convenient to have it take care of that concern. I guess this code might have to block on creating a new connection but that seems fine.